src/models/binary_graph_operations.py

```diff
@@ -0,0 +1,118 @@
+"""Contains tools for binary operations between GeoGraph objects"""
+from typing import List
+from numpy import ndarray
+from src.models.polygon_utils import (
+    connect_with_interior,
+    connect_with_interior_or_edge,
+    connect_with_interior_or_edge_or_corner,
+    connect_with_interior_bulk,
+    connect_with_interior_or_edge_bulk,
+    connect_with_interior_or_edge_or_corner_bulk,
+)
+
+# For switching identifiction mode in `identify_node`
+_SPATIAL_IDENTIFICATION_FUNCTION = {
+    "corner": connect_with_interior_or_edge_or_corner,
+    "edge": connect_with_interior_or_edge,
+    "interior": connect_with_interior,
+}
+_BULK_SPATIAL_IDENTIFICATION_FUNCTION = {
+    "corner": connect_with_interior_or_edge_or_corner_bulk,
+    "edge": connect_with_interior_or_edge_bulk,
+    "interior": connect_with_interior_bulk,
+}
+
+
+def identify_node(node: dict, other_graph: "GeoGraph", mode: str = "corner") -> ndarray:
+    """
+    Return list of all node ids in `other_graph` which identify with the given `node`.
+
+    Args:
+        node (dict): The node for which to find nodes in `other_graphs` that can be
+            identified with `node`.
+        other_graph (GeoGraph): The GeoGraph object in which to search for
+            identifications
+        mode (str, optional): Must be one of `corner`, `edge` or `interior`. Defaults
+            to "corner".
+            The different modes correspond to different rules for identification:
+            - corner: Polygons of the same `class_label` which overlap, touch in their
+                edges or corners will be identified with each other. (fastest)
+            - edge: Polygons of the same `class_label` which overlap or touch in their
+                edges will be identified with each other.
+            - interior: Polygons of the same `class_label` which overlap will be
+                identified with each other. Touching corners or edges are not counted.
+
+    Returns:
+        np.ndarray: List of node ids in `other_graph` which identify with `node`.
+    """
+
+    # Mode switch
+    assert mode in ["corner", "edge", "interior"]
+    have_valid_overlap = _BULK_SPATIAL_IDENTIFICATION_FUNCTION[mode]
+
+    # Extract relevant node elements for legibility
+    poly = node["geometry"]
+    label = node["class_label"]
+
+    # Get potential candidates for overlap
+    candidate_ids = other_graph.rtree.query(poly, sort=True)
+    # Filter candidates according to the same class label
+    # fmt: off
+    candidate_ids = candidate_ids[
+        other_graph._class_label(candidate_ids) == label  # pylint: disable=protected-access
+    ]
+    # Filter candidates accroding to correct spatial overlap
+    # fmt: on
+    candidate_ids = candidate_ids[
+        have_valid_overlap(
+            poly,
+            other_graph._geometry(candidate_ids),  # pylint: disable=protected-access
+        )
+    ]
+
```

Comment:
Intuitively it should be slightly faster to combine these two filters into one, right?

Reply:
Yes, I also think that if numpy does short-circuit evaluation on these things this should be faster.

Do you know how numpy handles these type of cases?
Concretely:
Case `select_from_array[np.logical_or(condition_array1, condition_array2)]`
Does it first evaluate both `condition_array1` and `condition_array2` in the slice `[ ... ]` and then `or` the conditions (in which case it'd probably be slower bc we would calculate the geometry overlaps for shapes which won't agree in class label).
Or does it calculate the first element of `condition_array1` and then short-circuit decide if that element of `condition_array2` even needs to be calculated? (in which case I think it should be slightly faster)

Reply:
I'm not sure about `np.logical_or` but if you use base python `or` and a generator to select then it should short-circuit and may be faster - there are some interesting timing results here on selecting from boolean arrays: https://stackoverflow.com/questions/58422690/filtering-a-numpy-array-what-is-the-best-approach